### ISSUE TITLE
Any plan to support OpenSSL?

### ISSUE THREAD

**Issue by cang-mang** [litespeedtech/lsquic#113]
Currently, lsquic only support BoringSSL. But in many phone devices, app-packet size is very important, and the old version of the app often takes OpenSSL. Now we add lsquic to the app, then we must add BoringSSL to it, too. So the size of the app-packet becomes too big. In some other cases, the old version of the app takes a static library of OpenSSL, and when we add lsquic with BoringSSL to the app, if the BoringSSL is a static library, then many symbols conflict; else if the BoringSSL is a dynamic library, then at run-time the app crashes. So, can you develop a branch version of supporting OpenSSL? Thanks very much.

**Comment by dtikhonov** [comment]
Yes, supporting OpenSSL would be very nice.  There are two obstacles to it:
1. It is not high on our priority list.
1. OpenSSL does not support QUIC yet.  (There is a patched version, but it's unofficial).

Note that BoringSSL can be compiled where all exported symbols are prefixed with a custom prefix.  That should allow you to use both OpenSSL and BoringSSL in your application.

**Comment by robmang** [comment]
How would we build lsquic to support boringssl with prefixed exported symbols?
Would it be up to us to rename them in lsquic?
Thanks

**Comment by dtikhonov** [comment]
No, you would not need to change any code.  See BoringSSL documentation.

**Comment by jazune** [comment]
> Yes, supporting OpenSSL would be very nice. There are two obstacles to it:
> 
> 1. It is not high on our priority list.
> 2. OpenSSL does not support QUIC yet.  (There is a patched version, but it's unofficial).
> 
> Note that BoringSSL can be compiled where all exported symbols are prefixed with a custom prefix. That should allow you to use both OpenSSL and BoringSSL in your application.

hi，@dtikhonov，  any play to suport openssl? If not, I wll do this. 
my plan is:
1，use openssl QUIC patch to suppot quic sdk.
2，in lsquic, replace AEAD api in boringssl with evp api which both exist in boringssl and openssl.
3，replace boringssl with openssl.

for lsquic，the main change is replacing AEAD  with evp。

**Comment by dtikhonov** [comment]
That would be super if you did it! 👍 💯 

I think this would require writing a set of function pointers: one for OpenSSL and one for BoringSSL.  But let's see what you come up with.  Of course, if you simply _replace_ BoringSSL with OpenSSL, we won't be able to merge it to mainline.

**Comment by jazune** [comment]
I plan only change lsquic code that will be suitable for both boringssl and openssl with quic sdk patch.
So, it won't effect current compile. And If one want to use openssl, it can also work.

**Comment by dtikhonov** [comment]
Great, I am looking forward to it!

**Comment by brjoha** [comment]
Has pluggable SSL percolated up the priority list at all?

**Comment by gegles** [comment]
Any update on this? BoringSSL is one of the reason we can't use this in our system at this point...

Thanks!

**Comment by alexpokotilo** [comment]
I've spent some time to build lsquic with prefix so here is my instruction. I need to add prefix as my project already has openssl as dependency.

1)First we need to build boringssl with prefixes. This is very strange why boringssl doesn't make this process done by just command line directive, but it is how it is. Please check boringssl docs about  DBORINGSSL_PREFIX for details.

I just followed these steps and generated symbol file for crypto lib and ssl lib and combined these files into crypto_ssl.txt
You can download crypto_ssl.txt generated from current master [crypto_ssl.txt](https://github.com/user-attachments/files/15911137/crypto_ssl.txt)
If you find problems during linking just regenerate crypto_ssl.txt for current version of boringssl

2) build boringssl with prefix
```
cmake -DBUILD_SHARED_LIBS=0 -DCMAKE_BUILD_TYPE=Release -DBORINGSSL_PREFIX=LSQUICK -DBORINGSSL_PREFIX_SYMBOLS=crypto_ssl.txt .
m
make -j 12 crypto ssl
cp symbol_prefix_include/* include/

```

without "cp symbol_prefix_include/* include/" lsquic will failed on build as symbol_prefix_include directory not added my lsquic CMakeFile into PATH


3) Build lsquic with prefix support
```
cmake -DLSQUIC_SHARED_LIB=OFF -DLSQUIC_BIN=OFF -DBORINGSSL_DIR=$BORINGSSL -DLSQUIC_TESTS=OFF -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS="-DBORINGSSL_PREFIX=LSQUICK" .
m
make -j 12
```

boringssl don't add BORINGSSL_PREFIX into generated headers. So you have to add BORINGSSL_PREFIX define manually to hint lsquic to build with prefix support. lsquic includes boringssl' headers, but to to be built with prefix you have to specify BORINGSSL_PREFIX during lsquic building again.

**Comment by topmanbest** [comment]
> 我花了一些时间用前缀构建 lsquic，所以这是我的说明。我需要添加前缀，因为我的项目已经有 openssl 作为依赖项。
> 
> 1)首先，我们需要使用前缀构建 boresssl。很奇怪，为什么 boresssl 不通过命令行指令完成此过程，但事实就是如此。请查看 boresssl 文档中有关 DBORINGSSL_PREFIX 的详细信息。
> 
> 我只是按照这些步骤生成了 crypto lib 和 ssl lib 的符号文件，并将这些文件合并到 crypto_ssl.txt 中。 您可以下载从当前主[crypto_ssl.txt](https://github.com/user-attachments/files/15911137/crypto_ssl.txt)生成的 crypto_ssl.txt。 如果您在链接过程中发现问题，只需为当前版本的 boressl 重新生成 crypto_ssl.txt。
> 
> 2. 使用前缀构建 boresssl
> 
> ```
> cmake -DBUILD_SHARED_LIBS=0 -DCMAKE_BUILD_TYPE=Release -DBORINGSSL_PREFIX=LSQUICK -DBORINGSSL_PREFIX_SYMBOLS=crypto_ssl.txt .
> make -j 12 crypto ssl
> cp symbol_prefix_include/* include/
> ```
> 
> 如果没有“cp symbol_prefix_include/* include/”，lsquic 将在构建时失败，因为 lsquic CMakeFile 未将 symbol_prefix_include 目录添加到 PATH 中
> 
> 3. 构建支持前缀的 lsquic
> 
> ```
> cmake -DLSQUIC_SHARED_LIB=OFF -DLSQUIC_BIN=OFF -DBORINGSSL_DIR=$BORINGSSL -DLSQUIC_TESTS=OFF -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_FLAGS="-DBORINGSSL_PREFIX=LSQUICK" .
> 
> make -j 12
> ```
> 
> boressl 不会将 BORINGSSL_PREFIX 添加到生成的标头中。因此，您必须手动添加 BORINGSSL_PREFIX 定义以提示 lsquic 使用前缀支持进行构建。lsquic 包含 boressl 的标头，但要使用前缀进行构建，您必须在 lsquic 构建期间再次指定 BORINGSSL_PREFIX。

非常感谢编译说明， thanks your answer。

**Comment by topmanbest** [comment]
@alexpokotilo   非常感谢你的编译说明。 thanks for your answer

**Comment by surpass007** [comment]
currently , openssl already supported QUIC , any plan to support OpenSSL?  @dtikhonov

**Comment by pkropachev** [comment]
@andrewkdinh, @nhorman, Andrew, Neil, could you please help to support OpenSSL in lsquic?

**Comment by nhorman** [comment]
We can discuss here and let you know if it can fit in our schedule.

**Comment by nhorman** [comment]
Hey all-
     So I spent the last two days taking a look at, and trying to start an update of lsquic to use openssl as an alternative to boringssl, and its going to be a pretty huge undertaking.  I've got a tree here:
https://github.com/nhorman/lsquic/tree/openssl

Fair warning, its unfinished and complete mess at the moment.

Please find below an enumeration of what I've done so far, and what still needs to be done

1) I modified the cmake file to detect if you are building against boringssl or openssl, and set a variable accordingly so we can do conditional compilation unit building, similarly to how  msquic and ngtcp2 enable support for multiple TLS backends.  I also renamed the BORINGSSL variables to OPENSSL, which isn't at all needed, but helps me with visual clarity while I'm trying to make sense of the code base, which I'm currently fairly unfamiliar with.

2) I started modifying call sites that use HDKF_extract and HKDF_expand to be conditionally compiled using HAVE_BORINGSSL vs HAVE_OPENSSL as defined in (1), so I can do separate implementations in those call sites using the openssl EVP_KDF api, as the boringssl HKDF api doesn't exist outside of boringssl.  I'm doing conditional compilation for now so that I can build for either implementation and compare behavior (once openssl is completely building of course, which I'm still some way away from).

3) Similar to (2) I did the same thing with the AEAD api from boring, as thats specific to boring, and we would need to use the EVP_CIPHER api here instead.  Also, much like (2), I'm fairly sure that I've not done the data marshalling correctly, so some debugging would be needed here once everything is building to sort out all the different use cases.  This and (2) took up most of the last few days.

4) The big one.  I started bifurcating lsquic_enc_sess_ietf.c to lsquic_enc_sess_ietf_boringssl.c and lsquic_enc_sess_ietf_openssl.c as the quic interface between the two implementations is completely different, so these files are going to diverge significantly.

I'm sorry, but given how spread out the call sites are for (2) and (3), and how much alteration is needed for (4), this is likely going to take several weeks to complete, and I'm not sure I have the time to dedicate to that at the moment.

If you're interested in getting support for openssl in lsquic however, I might suggest a different approach.  If you're interested , we could just incorporate some of the CMakeLists.txt changes I made to allow detection of which quic api is in use (boring vs openssl).  I could do that for you early next week.  That would still break on openssl of course, but it would clear the path to allowing builds with detection of the library in use.  From there we could start investigating on a case by case basis, the use of boringssl-specific calls, and replacing them with call sequences that will (hopefully) be compatible between both libraries.  Breaking it up like that, would allow for multiple people to start looking at the problem, and let us do comparative tests for each call use of HDKF_[expand|extract], and EVP_AEAD_CTX_[open|seal], testing as we go.  Once thats done, the last big bit (4) would be a much smaller lift, and would follow the same patterns that we used in msquic and ngtcp2.  

Thoughts?

**Comment by nhorman** [comment]
PR to update CMakeLists.txt to support multiple TLS backends submitted here:
https://github.com/litespeedtech/lsquic/pull/556

**Comment by pkropachev** [comment]
@nhorman, Neil, do you need any kind help in next steps? If so, could you please describe them? I suppose we can try to help for some simple cases.

**Comment by nhorman** [comment]
@pkropachev yes please.  With #556 merged, we're now free to move on addressing API inconsistencies.

One of the issues noted [here](https://github.com/litespeedtech/lsquic/issues/113#issuecomment-3169190868) is that lsquic has a number of call sites that use libcrypto api calls that are unique to boringssl.  The goal I think should be to modify those to use APIs that are common (or as common as possible) so that we can build with either boringssl or openssl.  For instance, the sha256() function in lsquic_crypto.c uses the boringssl SHA256 api call, which doesn't exist in openssl.  I would convert that to using the EVP_DigestInit/EVP_DigestUpdate/EVP_DigestFinal api, which is common to both libraries.  There are several such cases, which need to be individual addressed.

The general development cycle I use to find/address these is:

1) Configure lsquic to build with openssl (cmake -DLIBSSL_DIR=/path/to/openssl/build .)
2) Attempt to build lsquic, something will break (like the compilation of the above sha256 function)
3) Modify the broken function to use common boringssl/openssl api calls, using the HAVE_[OPENSSL|BORINGSSL] macros defined in #556 if necessary. 
4) Reconfigure lsquic with modifications from (3) against boringssl (cmake -DLIBSSL_DIR=/path/to/boringssl .
5) Build and run make tests to ensure that the build functions properly
6) Commit changes and open a PR
7) Lather, rinse, repeat

There will be a large number of these types of issues, so the progress will be a bit slow, but on the upside, they are all fairly isolated, and can be addressed individually and in parallel.  So if anyone has any free cycles, you can try the above steps, and address a specific api misuse.  I'll do the same of course, and eventually, we will address all of them.

Well, all of them except one, the use of SSL_provide_quic_data.  This is the boringssl hook for 3rd party quic stacks.  Openssl uses a significantly different api which will have to be addressed at the end.  I've done that for msquic and ngtcp2 however, and can handle that, once we fix all the other api inconsistencies above.

Thanks!
